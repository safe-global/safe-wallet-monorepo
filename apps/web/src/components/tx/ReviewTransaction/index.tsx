import { useContext } from 'react'
import { SafeTxContext } from '@/components/tx-flow/SafeTxProvider'
import ReviewTransactionSkeleton from './ReviewTransactionSkeleton'
import useTxDetails from '@/hooks/useTxDetails'
import useTxPreview from '../confirmation-views/useTxPreview'
import type { ReviewTransactionContentProps } from './ReviewTransactionContent'
import ReviewTransactionContent from './ReviewTransactionContent'

const ReviewTransaction = (props: ReviewTransactionContentProps) => {
  const { safeTx, safeTxError } = useContext(SafeTxContext)
  const [txDetails, , txDetailsLoading] = useTxDetails(props.txId)
  const [txPreview, , txPreviewLoading] = useTxPreview(safeTx?.data, undefined, props.txId)

  if ((!safeTx && !safeTxError) || txDetailsLoading || txPreviewLoading) {
    return <ReviewTransactionSkeleton />
  }

  return (
    <ReviewTransactionContent
      {...props}
      isCreation={!props.txId}
      txId={props.txId}
      txDetails={txDetails}
      txPreview={txPreview}
    >
      {props.children}
    </ReviewTransactionContent>
  )
}

export default ReviewTransaction
