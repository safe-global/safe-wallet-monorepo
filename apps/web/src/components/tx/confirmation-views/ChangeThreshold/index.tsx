import { Box, Divider, Typography } from '@mui/material'
import React from 'react'

import commonCss from '@/components/tx-flow/common/styles.module.css'
import useSafeInfo from '@/hooks/useSafeInfo'
import { ChangeSignerSetupWarning } from '@/features/multichain/components/SignerSetupWarning/ChangeSignerSetupWarning'
import { type TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import { isChangeThresholdView } from '../utils'
import { maybePlural } from '@safe-global/utils/utils/formatters'

interface ChangeThresholdProps {
  txInfo?: TransactionDetails['txInfo']
}

function ChangeThreshold({ txInfo }: ChangeThresholdProps) {
  const { safe } = useSafeInfo()
  const threshold = txInfo && isChangeThresholdView(txInfo) && txInfo.settingsInfo?.threshold

  return (
    <>
      <ChangeSignerSetupWarning />

      <div>
        <Typography variant="body2" color="text.secondary" mb={0.5}>
          Any transaction will require the confirmation of:
        </Typography>

        <Typography aria-label="threshold">
          <b>{threshold}</b> out of{' '}
          <b>
            {safe.owners.length} signer{maybePlural(safe.owners)}
          </b>
        </Typography>
      </div>
      <Box
        sx={{
          my: 1,
        }}
      >
        <Divider className={commonCss.nestedDivider} />
      </Box>
    </>
  )
}

export default ChangeThreshold
