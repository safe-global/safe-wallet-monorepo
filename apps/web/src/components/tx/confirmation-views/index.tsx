import type { TransactionPreview } from '@safe-global/safe-gateway-typescript-sdk'
import { type TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import type { SafeTransaction } from '@safe-global/safe-core-sdk-types'
import {
  isAnyStakingTxInfo,
  isCustomTxInfo,
  isExecTxData,
  isOnChainConfirmationTxData,
  isOnChainSignMessageTxData,
  isSafeMigrationTxData,
  isSafeUpdateTxData,
  isSwapOrderTxInfo,
  isTwapOrderTxInfo,
  isVaultDepositTxInfo,
  isVaultRedeemTxInfo,
} from '@/utils/transaction-guards'
import { type ReactNode, useContext, useMemo, useRef, useState, useEffect } from 'react'
import type { NarrowConfirmationViewProps } from './types'
import SettingsChange from './SettingsChange'
import ChangeThreshold from './ChangeThreshold'
import BatchTransactions from './BatchTransactions'
import { TxModalContext } from '@/components/tx-flow'
import { isSettingsChangeView, isChangeThresholdView, isConfirmBatchView } from './utils'
import { OnChainConfirmation } from '@/components/transactions/TxDetails/TxData/NestedTransaction/OnChainConfirmation'
import { ExecTransaction } from '@/components/transactions/TxDetails/TxData/NestedTransaction/ExecTransaction'
import { type ReactElement } from 'react'
import SwapOrder from './SwapOrder'
import StakingTx from './StakingTx'
import UpdateSafe from './UpdateSafe'
import { MigrateToL2Information } from './MigrateToL2Information'
import { NestedSafeCreation } from './NestedSafeCreation'
import { isNestedSafeCreation } from '@/utils/nested-safes'
import VaultDepositConfirmation from 'src/features/earn/components/VaultDepositConfirmation'
import VaultRedeemConfirmation from '@/features/earn/components/VaultRedeemConfirmation'
import Summary from '@/components/transactions/TxDetails/Summary'
import TxData from '@/components/transactions/TxDetails/TxData'
import { isMultiSendCalldata } from '@/utils/transaction-calldata'
import useChainId from '@/hooks/useChainId'
import { TransactionWarnings } from '../TransactionWarnings'
import { Box } from '@mui/material'
import DecodedData from '@/components/transactions/TxDetails/TxData/DecodedData'

type ConfirmationViewProps = {
  txDetails?: TransactionDetails
  txPreview?: TransactionPreview
  safeTx?: SafeTransaction
  isBatch?: boolean
  isApproval?: boolean
  isCreation?: boolean
  children?: ReactNode
}

const getConfirmationViewComponent = ({
  txInfo,
  txData,
  txFlow,
}: NarrowConfirmationViewProps & { txFlow?: ReactElement }) => {
  if (isChangeThresholdView(txInfo)) return <ChangeThreshold txInfo={txInfo} />

  if (isConfirmBatchView(txFlow)) return <BatchTransactions />

  if (isSettingsChangeView(txInfo)) return <SettingsChange txInfo={txInfo} />

  if (isOnChainConfirmationTxData(txData)) return <OnChainConfirmation data={txData} isConfirmationView />

  if (isExecTxData(txData)) return <ExecTransaction data={txData} isConfirmationView />

  if (isSwapOrderTxInfo(txInfo) || isTwapOrderTxInfo(txInfo)) return <SwapOrder txInfo={txInfo} txData={txData} />

  if (isAnyStakingTxInfo(txInfo)) return <StakingTx txInfo={txInfo} />

  // @ts-expect-error TODO: Fix these if there is time
  if (isVaultDepositTxInfo(txInfo)) return <VaultDepositConfirmation txInfo={txInfo} />

  // @ts-expect-error TODO: Fix these if there is time
  if (isVaultRedeemTxInfo(txInfo)) return <VaultRedeemConfirmation txInfo={txInfo} />

  if (isCustomTxInfo(txInfo) && isSafeUpdateTxData(txData)) return <UpdateSafe txData={txData} />

  if (isCustomTxInfo(txInfo) && isSafeMigrationTxData(txData)) {
    return <MigrateToL2Information variant="queue" txData={txData} />
  }

  if (isCustomTxInfo(txInfo) && txData && isNestedSafeCreation(txData)) {
    return <NestedSafeCreation txData={txData} />
  }

  return null
}

const ConfirmationView = ({ safeTx, txPreview, txDetails, ...props }: ConfirmationViewProps) => {
  const { txFlow } = useContext(TxModalContext)
  const details = txDetails ?? txPreview
  const chainId = useChainId()

  // Used to check if the decoded data was rendered inside the TxData component
  // If it was, we hide the decoded data in the Summary to avoid showing it twice
  const decodedDataRef = useRef(null)
  const [isDecodedDataVisible, setIsDecodedDataVisible] = useState(false)

  useEffect(() => {
    // If decodedDataRef.current is not null, the decoded data was rendered inside the TxData component
    setIsDecodedDataVisible(!!decodedDataRef.current)
  }, [])

  const ConfirmationViewComponent = useMemo(() => {
    return details
      ? getConfirmationViewComponent({
          txInfo: details.txInfo,
          txData: details.txData,
          txFlow,
        })
      : undefined
  }, [details, txFlow])

  const showTxDetails =
    details !== undefined &&
    !isMultiSendCalldata(details.txData?.hexData ?? '0x') &&
    !isOnChainSignMessageTxData(details?.txData, chainId)

  return (
    <>
      <TransactionWarnings txData={details?.txData} />
      {ConfirmationViewComponent ||
        (details && showTxDetails && (
          <TxData txData={details?.txData} txInfo={details?.txInfo} txDetails={txDetails} imitation={false} trusted>
            <Box ref={decodedDataRef}>
              <DecodedData
                txData={details.txData}
                toInfo={isCustomTxInfo(details.txInfo) ? details.txInfo.to : details.txData?.to}
              />
            </Box>
          </TxData>
        ))}

      {props.children}

      <Summary
        safeTxData={safeTx?.data}
        txDetails={txDetails}
        txData={details?.txData}
        txInfo={details?.txInfo}
        showDecodedData={!isDecodedDataVisible}
      />
    </>
  )
}

export default ConfirmationView
