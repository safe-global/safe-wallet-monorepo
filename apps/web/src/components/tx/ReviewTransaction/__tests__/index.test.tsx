import ReviewTransaction from '../index'
import { render } from '@/tests/test-utils'
import type { TransactionPreview } from '@safe-global/safe-gateway-typescript-sdk'
import type { SafeTxContextParams } from '@/components/tx-flow/SafeTxProvider'
import { SafeTxContext } from '@/components/tx-flow/SafeTxProvider'
import { createSafeTx } from '@/tests/builders/safeTx'
import * as useTxPreviewHooks from '@/components/tx/confirmation-views/useTxPreview'

describe('ReviewTransaction', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should display a loading component', () => {
    const { container } = render(<ReviewTransaction onSubmit={jest.fn()} />)

    expect(container).toMatchSnapshot()
  })

  it('should display a confirmation screen', async () => {
    jest.spyOn(useTxPreviewHooks, 'default').mockReturnValue([
      {
        txInfo: {},
        txData: {},
      } as TransactionPreview,
      undefined,
      false,
    ])

    const { container, getByTestId } = render(
      <SafeTxContext.Provider
        value={
          {
            safeTx: createSafeTx(),
          } as SafeTxContextParams
        }
      >
        <ReviewTransaction onSubmit={jest.fn()} />
      </SafeTxContext.Provider>,
    )

    expect(getByTestId('continue-sign-btn')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should display an error screen', async () => {
    jest
      .spyOn(useTxPreviewHooks, 'default')
      .mockReturnValue([undefined, new Error('This is a mock error message'), false])

    const { container } = render(
      <SafeTxContext.Provider
        value={
          {
            safeTx: createSafeTx(),
          } as SafeTxContextParams
        }
      >
        <ReviewTransaction onSubmit={jest.fn()} />
      </SafeTxContext.Provider>,
    )

    expect(container.querySelector('continue-sign-btn')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
