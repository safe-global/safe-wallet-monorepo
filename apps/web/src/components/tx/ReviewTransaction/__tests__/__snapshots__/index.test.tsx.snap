// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewTransaction should display a confirmation screen 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1ve84rd-MuiPaper-root-MuiAccordion-root"
            colorlevel="info"
            style="--Paper-shadow: none;"
          >
            <h3
              class="MuiAccordion-heading css-cy7rkm-MuiAccordion-heading"
            >
              <button
                aria-expanded="false"
                class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters accordion css-10sjung-MuiButtonBase-root-MuiAccordionSummary-root"
                data-testid="decoded-tx-summary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-1r0e0ir-MuiAccordionSummary-content"
                >
                  <div
                    class="MuiStack-root css-1bujbuo-MuiStack-root"
                  >
                    <div
                      class="MuiBox-root css-0"
                    >
                      Advanced details
                      <span
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <mock-icon
                          aria-hidden=""
                          class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-17ibv3e-MuiSvgIcon-root"
                          focusable="false"
                        />
                      </span>
                    </div>
                  </div>
                </span>
                <span
                  class="MuiAccordionSummary-expandIconWrapper css-1wqf3nl-MuiAccordionSummary-expandIconWrapper"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1dhtbeh-MuiSvgIcon-root"
                    data-testid="ExpandMoreIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    />
                  </svg>
                </span>
              </button>
            </h3>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
                >
                  <div
                    class="MuiAccordion-region"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-w74p4c-MuiAccordionDetails-root"
                      data-testid="decoded-tx-details"
                    >
                      <div
                        class="MuiStack-root css-14g2pqt-MuiStack-root"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-111k8jw-MuiTypography-root"
                        >
                          Transaction data
                        </p>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                            >
                              to:
                            </p>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            <div
                              class="container"
                            >
                              <div
                                class="MuiBox-root css-1lchl8k"
                              >
                                <div
                                  class="addressContainer"
                                >
                                  <div
                                    class="MuiBox-root css-b5p5gz"
                                  >
                                    <span
                                      aria-label="Copy to clipboard"
                                      class=""
                                      data-mui-internal-clone-element="true"
                                      style="cursor: pointer;"
                                    >
                                      <span>
                                        0x0000000000000000000000000000000000000000
                                      </span>
                                    </span>
                                  </div>
                                  <span
                                    aria-label="Copy to clipboard"
                                    class=""
                                    data-mui-internal-clone-element="true"
                                    style="cursor: pointer;"
                                  >
                                    <button
                                      aria-label="Copy to clipboard"
                                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-xyrzjn-MuiButtonBase-root-MuiIconButton-root"
                                      tabindex="0"
                                      type="button"
                                    >
                                      <mock-icon
                                        aria-hidden=""
                                        class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                                        data-testid="copy-btn-icon"
                                        focusable="false"
                                      />
                                    </button>
                                  </span>
                                  <div
                                    class="MuiBox-root css-yjghm1"
                                  />
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                            >
                              value:
                            </p>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-mxvcov-MuiTypography-root"
                            >
                              0x0
                            </p>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                            >
                              data:
                            </p>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            <div
                              class="encodedData MuiBox-root css-0"
                              data-testid="tx-hexData"
                            >
                              <span
                                aria-label="Copy to clipboard"
                                class=""
                                data-mui-internal-clone-element="true"
                                style="cursor: pointer;"
                              >
                                <button
                                  aria-label="Copy to clipboard"
                                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-xyrzjn-MuiButtonBase-root-MuiIconButton-root"
                                  tabindex="0"
                                  type="button"
                                >
                                  <mock-icon
                                    aria-hidden=""
                                    class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                                    data-testid="copy-btn-icon"
                                    focusable="false"
                                  />
                                </button>
                              </span>
                              0x
                               
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                            >
                              operation:
                            </p>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            0 (call)
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                            >
                              nonce:
                            </p>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            100
                          </div>
                        </div>
                        <div
                          class="MuiBox-root css-kz5wc9"
                        />
                        <p
                          class="MuiTypography-root MuiTypography-body1 css-111k8jw-MuiTypography-root"
                        >
                          Transaction hashes
                        </p>
                        <div
                          class="MuiStack-root css-hwnj0i-MuiStack-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                              data-testid="tx-row-title"
                              style="word-break: break-word;"
                            >
                              <p
                                class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                              >
                                Domain hash:
                              </p>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                              data-testid="tx-data-row"
                            >
                              <div
                                class="encodedData MuiBox-root css-0"
                                data-testid="tx-hexData"
                              >
                                <span
                                  aria-label="Copy to clipboard"
                                  class=""
                                  data-mui-internal-clone-element="true"
                                  style="cursor: pointer;"
                                >
                                  <button
                                    aria-label="Copy to clipboard"
                                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-xyrzjn-MuiButtonBase-root-MuiIconButton-root"
                                    tabindex="0"
                                    type="button"
                                  >
                                    <mock-icon
                                      aria-hidden=""
                                      class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                                      data-testid="copy-btn-icon"
                                      focusable="false"
                                    />
                                  </button>
                                </span>
                                 
                              </div>
                            </div>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-container css-1m1clne-MuiGrid-root"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item css-p1l0hx-MuiGrid-root"
                              data-testid="tx-row-title"
                              style="word-break: break-word;"
                            >
                              <p
                                class="MuiTypography-root MuiTypography-body1 css-shf88x-MuiTypography-root"
                              >
                                safeTxHash:
                              </p>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                              data-testid="tx-data-row"
                            >
                              <div
                                class="encodedData MuiBox-root css-0"
                                data-testid="tx-hexData"
                              >
                                <span
                                  aria-label="Copy to clipboard"
                                  class=""
                                  data-mui-internal-clone-element="true"
                                  style="cursor: pointer;"
                                >
                                  <button
                                    aria-label="Copy to clipboard"
                                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-xyrzjn-MuiButtonBase-root-MuiIconButton-root"
                                    tabindex="0"
                                    type="button"
                                  >
                                    <mock-icon
                                      aria-hidden=""
                                      class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                                      data-testid="copy-btn-icon"
                                      focusable="false"
                                    />
                                  </button>
                                </span>
                                 
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1whrnku-MuiTypography-root"
      >
        Transaction checks
      </h5>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <div
        class="wrapper"
      >
        <div
          class="icon sign"
        >
          <mock-icon
            aria-hidden=""
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-tqxw8e-MuiSvgIcon-root"
            focusable="false"
          />
        </div>
        <div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1mp5g96-MuiTypography-root"
          >
            confirm
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-17vdyq3-MuiTypography-root"
          >
            You're about to 
            create and 
            confirm
             this transaction.
          </p>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
      >
        <div
          class="MuiStack-root css-irtfmw-MuiStack-root"
        >
          <span
            aria-label="Please connect your wallet"
            class=""
            data-mui-internal-clone-element="true"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1hp2ejb-MuiButtonBase-root-MuiButton-root"
              data-testid="continue-sign-btn"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              Continue
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ReviewTransaction should display a loading component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <div
        class="MuiBox-root css-17luq05"
      >
        <div
          class="box MuiBox-root css-0"
        >
          <div
            class="rect rectTl"
          />
          <div
            class="rect rectTr"
          />
          <div
            class="rect rectBl"
          />
          <div
            class="rect rectBr"
          />
          <div
            class="rect rectCenter"
          />
          <svg
            version="1.1"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <filter
                id="gooey"
              >
                <fegaussianblur
                  in="SourceGraphic"
                  result="blur"
                  stdDeviation="3"
                />
                <fecolormatrix
                  in="blur"
                  mode="matrix"
                  result="goo"
                  values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 19 -9"
                />
                <fecomposite
                  in="SourceGraphic"
                  in2="goo"
                  operator="atop"
                />
              </filter>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ReviewTransaction should display an error screen 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1ve84rd-MuiPaper-root-MuiAccordion-root"
            colorlevel="info"
            style="--Paper-shadow: none;"
          >
            <h3
              class="MuiAccordion-heading css-cy7rkm-MuiAccordion-heading"
            >
              <button
                aria-expanded="false"
                class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters accordion css-10sjung-MuiButtonBase-root-MuiAccordionSummary-root"
                data-testid="decoded-tx-summary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-1r0e0ir-MuiAccordionSummary-content"
                >
                  <div
                    class="MuiStack-root css-1bujbuo-MuiStack-root"
                  >
                    <div
                      class="MuiBox-root css-0"
                    >
                      Advanced details
                      <span
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <mock-icon
                          aria-hidden=""
                          class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-17ibv3e-MuiSvgIcon-root"
                          focusable="false"
                        />
                      </span>
                    </div>
                  </div>
                </span>
                <span
                  class="MuiAccordionSummary-expandIconWrapper css-1wqf3nl-MuiAccordionSummary-expandIconWrapper"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1dhtbeh-MuiSvgIcon-root"
                    data-testid="ExpandMoreIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    />
                  </svg>
                </span>
              </button>
            </h3>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
                >
                  <div
                    class="MuiAccordion-region"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-w74p4c-MuiAccordionDetails-root"
                      data-testid="decoded-tx-details"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1whrnku-MuiTypography-root"
      >
        Transaction checks
      </h5>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-1w4z6qv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiCardContent-root cardContent css-1lt5qva-MuiCardContent-root"
      data-testid="card-content"
    >
      <div
        class="wrapper"
      >
        <div
          class="icon sign"
        >
          <mock-icon
            aria-hidden=""
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-tqxw8e-MuiSvgIcon-root"
            focusable="false"
          />
        </div>
        <div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1mp5g96-MuiTypography-root"
          >
            confirm
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-17vdyq3-MuiTypography-root"
          >
            You're about to 
            create and 
            confirm
             this transaction.
          </p>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
      >
        <div
          class="MuiStack-root css-irtfmw-MuiStack-root"
        >
          <span
            aria-label="Please connect your wallet"
            class=""
            data-mui-internal-clone-element="true"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1hp2ejb-MuiButtonBase-root-MuiButton-root"
              data-testid="continue-sign-btn"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              Continue
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
